Fix issue #15. MellonSetEnv Directives must be in same block.

briantward opened the pull request:
Rather than check environment variables with IF statements, expect them all to be set as a given.  
MellonSetEnv fails to work properly when there are multiple directives in separate code blocks (in this case IF blocks).  This seems to be a bug with mod_auth_mellon.  
Parsing this way is a work around for now.
RequestHeaders will only be set when the corresponding attribute contains a value anyways (using "env=VARIABLE_NAME" at the end of the line).

I also removed the fallback for setting the Remote-User to NAME_ID when not providing any option, which means REMOTE_USER_SAML_ATTRIBUTE is required.  NAME_ID is transient in mod_auth_mellon by default, which is probably a good design here anyways.  It allows mellon to define a unique user id before ever connecting to the IdP.  We never want to use it with respect to actually logging into OpenShift because it changes every time there is a new session.

Commit:
Fix issue #15. MellonSetEnv Directives must be in same block.

briantward commented:
@brenton @itewk any thoughts?

itewk commented:
@briantward what IdP did you test this with? Did you run through the different scenarios of what would happen if one of the environment variables isn't set, or if the IdP isn't sending one of the attributes? I would be curious if there are any error messages we could add to the debugging section of the doc so as to make troubleshooting missing something that is required easier. 

Unfortunately I do not have any test environments currently set up to test myself, but your explanation and changes on the whole make sense and "look" good to me.

briantward commented:
@itewk I tested with RH-SSO, with mappings for id, username, email, and fullname.  In every test case I used the same mappings from the IdP and did not change them.

  MELLON_id: <UID from RH-SSO>
  MELLON_id_0: <UID from RH-SSO>
  MELLON_email: user@email.com
  MELLON_email_0: user@email.com
  MELLON_fullname: Test
  MELLON_fullname_0: Test
  MELLON_username: test-user
  MELLON_username_0: test-user

Here are some notes of the tests:
1. I removed all environment variables other than REMOTE_USER_SAML_ATTRIBUTE set to username.  This worked perfectly fine.  But of course I did not get mappings for my other attributes.  This was expected.
2. I added other variables back to the dc and things stopped working.  I tested each one. I tested by removing the IF blocks specifically as well.  
3. Email always came through.  This seems obvious, because the variable mapping was one to one with no changes "email" from mellon was "email" from IdP.  I suspect MellonSetEnv was still not working from this context, but the assignment worked because Mellon sets all IdP user attributes to Mellon_$ATTR.  Likewise, if your IdP returned "user" instead of "username", I would expect the mapping to be automatic without a MellonSetEnv.  If you wanted to, you could just "require" all field values from the IdP match your chosen mellon environment variables.  Then you would not need any MellonSetEnv directives at all.  I did attempt to get a chosen environment variable to resolve similar to the below, but it failed to work:
%{Mellon_${REMOTE_USER_SAML_ATTRIBUTE}}
4. In all these test failures, where I should have been getting all four mappings assigned based on the conditions of the apache conf file, I recorded only "MellonSetEnv (envattr): 1 items" in the mellon diagnostics file.  This pretty much proved to me, that even though the conditions were valid, the MellonSetEnv directives were not working correctly.  
5. I removed all IF blocks and all MellonSetEnv directives worked fine.  

I'm working on a full test-suite deployer that I'll share shortly, that provides both RH-SSO and Mellon and updates the cluster configuration.

itewk commented:
@briantward want to make part of this PR updating the docs since there is currentl a warning in the docs that you can't really use those attributes because of this bug?

briantward commented:
@itewk good point. will do.

itewk commented:
@briantward ever get a chance to update the doc?

briantward commented:
@itewk Sorry got sidetracked with other more critical work.  I'll have some time to focus on this again now.

Commit:
Update the README to match findings on attribute variables.

briantward commented:
@itewk updated the README to reflect the knowledge around the attributes.  I'll have further updates on the merge with the other work.

itewk commented:
@brenton this looks good to me. You available to do a review and merge?

brenton commented:
Thanks again for all the help hardening this configuration.

briantward commented:
@itewk updated the README to reflect the knowledge around the attributes.  I'll have further updates on the merge with the other work.

briantward commented:
woops, ignore that last comment...
